File: ts/src/UuidGeneratorSDK.ts
// UuidGenerator Ts SDK

import { DecodeEntity } from './entity/DecodeEntity'
import { TimestampFirstEntity } from './entity/TimestampFirstEntity'
import { Version1Entity } from './entity/Version1Entity'
import { Version3Entity } from './entity/Version3Entity'
import { Version4Entity } from './entity/Version4Entity'
import { Version5Entity } from './entity/Version5Entity'

export type * from './UuidGeneratorTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { UuidGeneratorEntityBase } from './UuidGeneratorEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class UuidGeneratorSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _decode?: DecodeEntity

  // Idiomatic facade: `client.decode.list()` / `client.decode.load({ id })`.
  get decode(): DecodeEntity {
    return (this._decode ??= new DecodeEntity(this, undefined))
  }

  /** @deprecated Use `client.decode` instead. */
  Decode(data?: any) {
    const self = this
    return new DecodeEntity(self,data)
  }


  _timestamp_first?: TimestampFirstEntity

  // Idiomatic facade: `client.timestamp_first.list()` / `client.timestamp_first.load({ id })`.
  get timestamp_first(): TimestampFirstEntity {
    return (this._timestamp_first ??= new TimestampFirstEntity(this, undefined))
  }

  /** @deprecated Use `client.timestamp_first` instead. */
  TimestampFirst(data?: any) {
    const self = this
    return new TimestampFirstEntity(self,data)
  }


  _version_1?: Version1Entity

  // Idiomatic facade: `client.version_1.list()` / `client.version_1.load({ id })`.
  get version_1(): Version1Entity {
    return (this._version_1 ??= new Version1Entity(this, undefined))
  }

  /** @deprecated Use `client.version_1` instead. */
  Version1(data?: any) {
    const self = this
    return new Version1Entity(self,data)
  }


  _version_3?: Version3Entity

  // Idiomatic facade: `client.version_3.list()` / `client.version_3.load({ id })`.
  get version_3(): Version3Entity {
    return (this._version_3 ??= new Version3Entity(this, undefined))
  }

  /** @deprecated Use `client.version_3` instead. */
  Version3(data?: any) {
    const self = this
    return new Version3Entity(self,data)
  }


  _version_4?: Version4Entity

  // Idiomatic facade: `client.version_4.list()` / `client.version_4.load({ id })`.
  get version_4(): Version4Entity {
    return (this._version_4 ??= new Version4Entity(this, undefined))
  }

  /** @deprecated Use `client.version_4` instead. */
  Version4(data?: any) {
    const self = this
    return new Version4Entity(self,data)
  }


  _version_5?: Version5Entity

  // Idiomatic facade: `client.version_5.list()` / `client.version_5.load({ id })`.
  get version_5(): Version5Entity {
    return (this._version_5 ??= new Version5Entity(this, undefined))
  }

  /** @deprecated Use `client.version_5` instead. */
  Version5(data?: any) {
    const self = this
    return new Version5Entity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new UuidGeneratorSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return UuidGeneratorSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'UuidGenerator' }
  }

  toString() {
    return 'UuidGenerator ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = UuidGeneratorSDK


export {
  stdutil,

  BaseFeature,
  UuidGeneratorEntityBase,

  UuidGeneratorSDK,
  SDK,
}

File: go/core/uuid-generator_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/uuid-generator-sdk/go/utility/struct"
)

type UuidGeneratorSDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewUuidGeneratorSDK(options map[string]any) *UuidGeneratorSDK {
	sdk := &UuidGeneratorSDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features from config.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		for _, item := range vs.Items(featureOpts) {
			fname, _ := item[0].(string)
			fopts := ToMapAny(item[1])
			if fopts != nil {
				if active, ok := fopts["active"]; ok {
					if ab, ok := active.(bool); ok && ab {
						sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *UuidGeneratorSDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *UuidGeneratorSDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *UuidGeneratorSDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *UuidGeneratorSDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *UuidGeneratorSDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Decode returns a Decode entity bound to this client.
// Idiomatic usage: client.Decode(nil).List(nil, nil) or
// client.Decode(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *UuidGeneratorSDK) Decode(data map[string]any) UuidGeneratorEntity {
	return NewDecodeEntityFunc(sdk, data)
}


// TimestampFirst returns a TimestampFirst entity bound to this client.
// Idiomatic usage: client.TimestampFirst(nil).List(nil, nil) or
// client.TimestampFirst(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *UuidGeneratorSDK) TimestampFirst(data map[string]any) UuidGeneratorEntity {
	return NewTimestampFirstEntityFunc(sdk, data)
}


// Version1 returns a Version1 entity bound to this client.
// Idiomatic usage: client.Version1(nil).List(nil, nil) or
// client.Version1(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *UuidGeneratorSDK) Version1(data map[string]any) UuidGeneratorEntity {
	return NewVersion1EntityFunc(sdk, data)
}


// Version3 returns a Version3 entity bound to this client.
// Idiomatic usage: client.Version3(nil).List(nil, nil) or
// client.Version3(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *UuidGeneratorSDK) Version3(data map[string]any) UuidGeneratorEntity {
	return NewVersion3EntityFunc(sdk, data)
}


// Version4 returns a Version4 entity bound to this client.
// Idiomatic usage: client.Version4(nil).List(nil, nil) or
// client.Version4(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *UuidGeneratorSDK) Version4(data map[string]any) UuidGeneratorEntity {
	return NewVersion4EntityFunc(sdk, data)
}


// Version5 returns a Version5 entity bound to this client.
// Idiomatic usage: client.Version5(nil).List(nil, nil) or
// client.Version5(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *UuidGeneratorSDK) Version5(data map[string]any) UuidGeneratorEntity {
	return NewVersion5EntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *UuidGeneratorSDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewUuidGeneratorSDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}
